Parameterise base configs so they can be overridden.

jinnko opened the pull request:
The standard deployment behaviour doesn't change with this commit, however
it's now possible to override the default config values.  The main motivation for
this is to be able to recurse rDNS lookups, which requires removing the
'bogus-priv' config parameter.
- Introduce a new params class and migrate the defaults that existed in
  the static config files.
- Update the init, client and server classes to deploy their relevant
  config files from templates, taking their values from the new params
  class.
- Remove the old static config files.
- Update documentation to reflect the changes.

jinnko committed:
Parameterise base configs so they can be overridden.

The standard deployment behaviour doesn't change with this commit, however
it's now possible to override the default config values.  The main
motivation for this is to be able to recurse rDNS lookups, which requires removing the 'bogus-priv' config parameter.

- Introduce a new params class and migrate the defaults that existed in
  the static config files.
- Update the init, client and server classes to deploy their relevant
  config files from templates, taking their values from the new params
  class.
- Remove the old static config files.
- Update documentation to reflect the changes.

mattbostock commented on manifests/init.pp:
```diff
@@ -20,10 +20,20 @@
+#   Default config options for the server.
+#   Default:
+#     options:
+#       no-negcache: ''
```
This implies that `no-negcache` takes a string. Would a boolean be more suitable?

mattbostock replied:
Though using a boolean would have an impact on how you generate the template.

mattbostock commented on templates/dnsmasq/config_options.conf.erb:
```diff
@@ -0,0 +1,5 @@
+# Puppet managed options
```
"Options managed by Puppet" might be clearer.

mattbostock commented:
:+1: for the change, just some minor comments about the implementation.

queeno committed:
Sort the options in the template before building it

The options array in the config_options.conf.erb template should be
sorted before the file is generated. This allow the file to be produced
in a deterministic way, puppet to run idempotently and the subscribed
services not to be refreshed.

jinnko committed:
Merge pull request #1 from queeno/master

Sort the options in the template before building it

mattbostock commented:
@jinnko Are you able to address the comments in this PR? Thanks.

mattbostock commented:
Closing as this PR hasn't been updated in some time.
